### ISSUE TITLE
0.5 won't launch from app icon until xapp python3-xapp is installed

### ISSUE THREAD

**Issue by seraph** [getting-things-gnome/gtg#621]
After installing GTG 0.5 from flatpak/flathub on a fresh fedora 33 system:

1. Unable to launch the app using the app icon / gnome launcher
2. `flatpak run org.gnome.GTG -d` worked without xapps python3-xapp being installed
3. `sudo dnf install xapps python3-xapp` fixed launching via app icon

There were no error messages about xapp being missing unlike in issue #616 and it still worked with the `flatpak run -d` command which ... is strange.

Is it missing from the flatpak deps, or is it that xapps needs to be installed on the host system natively?

Thanks. Let me know if you require further details.

**Comment by diegogangl** [comment]
This might be a Gnome 40 platform change, because we don't have any dependencies to python-xapp. Looking at the API there doesn't seem to be anything we're using, even indirectly

**Comment by nekohayo** [comment]
This is weird, because I'm on Fedora 33 as well, I don't have python3-xapp installed, and GTG 0.5 (flatpak version) launches from my GNOME Shell's "dash-to-panel" extension dock without any problems, no matter how many times I shut it down and retry.

**Comment by seraph** [comment]
> This is weird, because I'm on Fedora 33 as well, I don't have python3-xapp installed, and GTG 0.5 (flatpak version) launches from my GNOME Shell's "dash-to-panel" extension dock without any problems, no matter how many times I shut it down and retry. Could it have something to do with some plugin @seraph might have activated?

Plug-in for GTG itself, or for GNOME Shell? If you mean GTG then it was a fresh install and I've never used the app before either so there would be no cruft left from before like plugin settings etc. As for GNOME Shell I just run stuff like app indicator support, workspace indicator and a couple of other things to add items to the top right systray area.

**Comment by fernandochiquito** [comment]
So I'm having the same problem.
For me when I run the command `flatpak run org.gnome.GTG -d` this is the outcome.

`GNOME keyring not found, passwords will be not stored after restarting GTG
Gdk-Message: 08:41:10.751: Unable to load hand2 from the cursor theme
Traceback (most recent call last):
  File "/app/bin/gtg", line 79, in <module>
    from GTG.gtk.application import Application
  File "/app/lib/python3.8/site-packages/GTG/gtk/application.py", line 28, in <module>
    from GTG.gtk.browser.main_window import MainWindow
  File "/app/lib/python3.8/site-packages/GTG/gtk/browser/main_window.py", line 35, in <module>
    from GTG.gtk.browser import quick_add
  File "/app/lib/python3.8/site-packages/GTG/gtk/browser/quick_add.py", line 27, in <module>
    from GTG.gtk.editor.taskview import TAG_REGEX
  File "/app/lib/python3.8/site-packages/GTG/gtk/editor/taskview.py", line 54, in <module>
    class TaskView(Gtk.TextView):
  File "/app/lib/python3.8/site-packages/GTG/gtk/editor/taskview.py", line 78, in TaskView
    CURSOR_HAND = Gdk.Cursor.new(Gdk.CursorType.HAND2)
TypeError: constructor returned NULL
`

So I tried to install xapps python3-xapp and I got the same outcome.

**Comment by diegogangl** [comment]
@fernandochiquito do you mean installing xapps fixed it? What distro/version are you running?

**Comment by santiagocezar** [comment]
The exact same error @fernandochiquito 
```
❯ flatpak run org.gnome.GTG -d
GNOME keyring not found, passwords will be not stored after restarting GTG
Gdk-Message: 11:52:13.791: Unable to load hand2 from the cursor theme
Traceback (most recent call last):
  File "/app/bin/gtg", line 79, in <module>
    from GTG.gtk.application import Application
  File "/app/lib/python3.8/site-packages/GTG/gtk/application.py", line 28, in <module>
    from GTG.gtk.browser.main_window import MainWindow
  File "/app/lib/python3.8/site-packages/GTG/gtk/browser/main_window.py", line 35, in <module>
    from GTG.gtk.browser import quick_add
  File "/app/lib/python3.8/site-packages/GTG/gtk/browser/quick_add.py", line 27, in <module>
    from GTG.gtk.editor.taskview import TAG_REGEX
  File "/app/lib/python3.8/site-packages/GTG/gtk/editor/taskview.py", line 54, in <module>
    class TaskView(Gtk.TextView):
  File "/app/lib/python3.8/site-packages/GTG/gtk/editor/taskview.py", line 78, in TaskView
    CURSOR_HAND = Gdk.Cursor.new(Gdk.CursorType.HAND2)
TypeError: constructor returned NULL
```

I'm running Fedora release 34 (Thirty Four), GTG 0.5, Flatpak 1.10.2, and I've installed xapps and python3-xapp (didn't change anything).

This is a problem with the flatpak, GTG works when compiling from master

**Comment by santiagocezar** [comment]
It turns out it does work if I use the default (Adwaita) cursor theme, I was using Bibata. previously.

I tried with other Flatpak apps and they seem to use a fallback cursor theme when not using Adwaita

- Adwaita:  
  ![Captura de pantalla de 2021-04-26 12-15-05](https://user-images.githubusercontent.com/48573316/116107203-34b5eb80-a689-11eb-8b0b-86a095564db6.png)
- Bibata:
  ![Captura de pantalla de 2021-04-26 12-14-33](https://user-images.githubusercontent.com/48573316/116107207-354e8200-a689-11eb-8056-245def5eb2a9.png)

**Comment by Neui** [comment]
> GTG works when compiling from master

Actually, the used function was deprecated and the replacement needed one more info, thus I had to move it from startup to when it is being needed, so if you just checked whenever GTG starts up you didn't really test whenever it still works.

To check in current master, open any task and move the cursor around in the text-part of the editor (uses "XTERM" cursor, not sure if it works there, your error is before it tries it out), and try to hover over a tag (uses the faulty "HAND2" one).

The Issue is technically at GTK/GDK since the documented deprecated function returns something it claims it never does (more technical: `Cursor.get()` can now return `None` because it uses `Cursor.get_from_display()` [under the hood](https://gitlab.gnome.org/GNOME/gtk/-/blob/gtk-3-24/gdk/gdkcursor.c#L185-202), but the documentation for the former one doesn't say so and trips up PyGObject).

However, on master it can work, but I am not sure it is the expected result when hovering over a tag (either a normal cursor when the "error" still exists (should print the log "Unable to load hand2 from the cursor theme" message), or the HAND2 if the issue is something with flatpak or something else, assuming you're running outside of flatpak)

**Comment by leio** [comment]
`Gdk.Cursor.new` (used in 0.5) and `Gdk.Cursor.new_for_display` (used in git master) both could fail and return null, in which case the set_cursor call shouldn't be done. But maybe we can figure a fallback cursor name first to try, if hand2 isn't there?

`gdk_cursor_new_from_name` GDK 3 docs document a recommended set of cursor names, and hand2 isn't amongst them, nor is xterm (used elsewhere then the stacktrace seen here). I think these have X11-specific fallbacks that won't be there when running on (native) Wayland.

Note that GdkCursorType docs also claim
> The recommended way to create cursors is to use gdk_cursor_new_from_name().

Whereas gdk_cursor_new suggests to use gdk_cursor_new_for_display like git master does now.

**Comment by Neui** [comment]
> I think these have X11-specific fallbacks that won't be there when running on (native) Wayland.

Good point, although it seems they have [some kind of mapping for Wayland](https://gitlab.gnome.org/GNOME/gtk/-/blob/gtk-3-24/gdk/wayland/gdkcursor-wayland.c#L403-405) (not entirely sure how exactly it works/what it contains), but even then I'd assume it uses it by name, and the cursor ships a cursor with the filename of [hand2](https://github.com/ful1e5/Bibata_Cursor/blob/main/svg/modern/static/hand2.svg). (I've never worked on cursor themes, so I am not totally sure how it works) The cursor theme works with GTG on X11 for me though.

Also, we should move to `gdk_cursor_new_from_name()` since GDK 4 (for when we migrate to GTK 4) dropped `gdk_cursor_new_for_display()`.

(Also I am not sure if this is a new issue, since it sounds like xapp has nothing to do with it? I don't even see where we explicitly even use xapp)

**Comment by santiagocezar** [comment]
> Actually, the used function was deprecated and the replacement needed one more info, thus I had to move it from startup to when it is being needed, so if you just checked whenever GTG starts up you didn't really test whenever it still works.

It does work, even when using Bibata, on the host system. The problem appears when running the Flatpak'd version of the app.

I cloned and modified the [flapak manifest file](https://github.com/flathub/org.gnome.GTG/blob/master/org.gnome.GTG.json) to build the latest commit and now the app does start, and it uses a fallback cursor like Epiphany does. when I hover any element that uses the HAND2 cursor it uses whatever was the previous cursor

![screenshot b](https://user-images.githubusercontent.com/48573316/116151268-c68a1c80-a6ba-11eb-98fd-7cf63af1a7db.png)

and on the terminal prints this when I move it over
```
Traceback (most recent call last):
  File "/app/lib/python3.8/site-packages/GTG/gtk/editor/taskview.py", line 616, in on_mouse_move
    cursor = Gdk.Cursor.new_for_display(window.get_display(),
TypeError: constructor returned NULL
```
Looks like the try except block is saving the app from crashing.

Anyways, I think this is more of a Flatpak issue, because after running `flatpak --user override org.gnome.GTG --filesystem=/home/$USER/.icons/:ro` the app works perfectly

> (Also I am not sure if this is a new issue, since it sounds like xapp has nothing to do with it? I don't even see where we explicitly even use xapp)

Looks like it wasn't related to xapp, but I saw that fernando had the exact same error that I got and decided to post here. Should I make a new issue?

**Comment by Neui** [comment]
>     TypeError: constructor returned NULL

Seems wrong to me, I've submitted an issue upstream: https://gitlab.gnome.org/GNOME/pygobject/-/issues/469

> Anyways, I think this is more of a Flatpak issue, because after running `flatpak --user override org.gnome.GTG --filesystem=/home/$USER/.icons/:ro` the app works perfectly

Ok, one part figured out. However, it seems using the XDG directories would be better, but...
I've found an [draft spec](https://freedesktop.org/wiki/DesktopThemeSpec/) about placing cursor themes in "$XDG_DATA_DIRS/cursors". I placed the theme folders at "~/.local/share/cursors", but gnome-tweaks didn't show them. The [icon spec](https://specifications.freedesktop.org/icon-theme-spec/latest/ar01s03.html) says "$XDG_DATA_DIRS/icons", but placing them in "~/.local/share/icons" shows up, but selecting doesn't seem to apply(?). "$HOME/.icons" is explicitly labelled as "backward compatibility", so likely flatpak isn't looking into it because of that. Probably I really need to place it into one of the `$XDG_DATA_DIRS` directories, that doesn't seem to include "~/.local/share"...

> I saw that fernando had the exact same error that I got and decided to post here, but after all I don't think it's a problem with xapp. Should I make a new issue?

I guessed so, and that is OK, but I think a new issue is better, since it doesn't seem to be anything about xapp (that we don't use), except installing xapp somehow magically fixes this.

**Comment by santiagocezar** [comment]
Moved to #647

**Comment by fernandochiquito** [comment]
> It turns out it does work if I use the default (Adwaita) cursor theme, I was using Bibata. previously.
> 
> I tried with other Flatpak apps and they seem to use a fallback cursor theme when not using Adwaita
> 
> * Adwaita:
>   ![Captura de pantalla de 2021-04-26 12-15-05](https://user-images.githubusercontent.com/48573316/116107203-34b5eb80-a689-11eb-8b0b-86a095564db6.png)
> * Bibata:
>   ![Captura de pantalla de 2021-04-26 12-14-33](https://user-images.githubusercontent.com/48573316/116107207-354e8200-a689-11eb-8056-245def5eb2a9.png)

So i just change my cursor to the default adwaita theme and now its working!
I did remove xapps python3-xapp form my system and launch the flatpak version and its just working...

**Comment by nekohayo** [comment]
Trying to make sense of what I'm seeing in this ticket,
* it seems the majority (or only actionable part) of the issue is about the cursors handling, which is potentially fixed by the fixfor issue #647, so closing as a duplicate of that one.
* I'm running the flatpak version in Fedora 33 and, without ever having installed python3-xapp, the app works, so again I'll presume this to be related to the user's cursors theme and the API use that got solved in PR 658...
